frb_codegen/src/config/opts.rs

```diff
@@ -29,17 +43,218 @@ pub struct Opts {
+    // for the below 2 fields:
+    // in single-block case, they should be `None`;
+    // in multi-blocks case, they should be paths to generated files with full directories
+    pub shared_rust_output_path: Option<String>,
+    pub shared_dart_output_path: Option<String>,
     pub keep_going: bool,
 }
 
 impl Opts {
-    pub fn get_ir_file(&self) -> ParserResult<IrFile> {
-        // info!("Phase: Parse source code to AST");
-        let source_rust_content = read_rust_file(&PathBuf::from(&self.rust_input_path));
-        let file_ast = syn::parse_file(&source_rust_content)?;
+    /// NOTE: for `Opts` for a regular API block, if `all_configs` is empty or with only 1 element,
+    /// it would return an `IrFile` with field of EMPTY `shared_types` for the regular block;
+    /// for `Opts` for an auto-generated shared API block, make sure `all_configs` has at least 2 items, each of which
+    /// is for the regular block. Otherwise, it would panic.
+    pub fn get_ir_file(&self, all_configs: &[Opts]) -> ParserResult<IrFile> {
+        let raw_ir_file = if !self.shared {
+            self.get_regular_ir_file(all_configs)?
+        } else {
+            self.get_shared_ir_file(all_configs)?
+        };
 
-        // info!("Phase: Parse AST to IR");
-        parser::parse(&source_rust_content, file_ast, &self.manifest_path)
+        log::debug!("Phase: Transform IR");
+        let ir_file = transformer::transform(raw_ir_file);
+
+        Ok(ir_file)
+    }
+
+    fn get_regular_ir_file(&self, all_configs: &[Opts]) -> ParserResult<IrFile> {
+        let mut ir_file = IR_FILE_MAP.with(|data| {
+            let mut ir_file_map = data.borrow_mut();
+            let ir_file = if let std::collections::hash_map::Entry::Vacant(e) =
+                ir_file_map.entry(self.block_index)
+            {
+                log::debug!("Phase: Parse source code to AST");
+                let source_rust_content = read_rust_file(&PathBuf::from(&self.rust_input_path));
+                let file_ast = syn::parse_file(&source_rust_content).unwrap();
+                log::debug!("Phase: Parse AST to IR");
+                let ir_file = parser::parse(
+                    &source_rust_content,
+                    file_ast,
+                    &self.manifest_path,
+                    self.block_index,
+                    self.shared,
+                    all_configs,
+                )
+                .unwrap();
+
+                e.insert(ir_file.clone());
+
+                ir_file
+            } else {
+                ir_file_map[&self.block_index].clone()
+            };
+
+            ir_file
+        });
+
+        // if this is a single block case or `all_configs` is empty,
+        // then there is no need to deal with funcs with
+        // extra API blocks or extra rust module files.
+        if all_configs.len() <= 1 {
+            return Ok(ir_file);
+        }
+
+        ir_file.fetch_shared_types_if_needed(all_configs);
```

Comment:
it would be great to make IrFile immutable - as we know, immutability is great and removes a lot of bugs and problems